### PR TITLE
TR rules improvements (function, context, model names and descriptions)

### PR DESCRIPTION
**Pull request by alicekwak**
TR rules improvements (function, context, model names and descriptions):

- TR rules for extracting model names and descriptions added
- More function rules and context rules added
- Cross sentence mention was added (to be used in combining function fragments)
- A method for model coreference resolution was added
- Expansion handler was edited to make adjustments for model description mentions
- Cosmos document, Json processor, and data loader were revised to filter out section headers

### CODE REVIEW

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceMention.scala
Comment (maxaalexeeva) [inline automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceMention.scala:25]: Hm, if you use this type of mention, we need to figure out serialization for it before we merge. I can help with that (I can do it or we can meet and talk about it). Also, are there any issues with the span of the mention? I guess if the arg text shows up correctly, the full mention span being somewhat off, that's alright. Let's discuss!

File: automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml
Comment (maxaalexeeva) [reply automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml:155]: We may want to consider writing some tests for model descriptions.

**Review by maxaalexeeva — CHANGES_REQUESTED** [review]
I did an initial pass. Will do another one after master is merged into this branch. Mentioning a few topics for us to discuss/work on before merging as well (cross-sent mention serialization, cross-sent mention text span, and adding a few tests for model descriptions)

File: automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/entities.yml
Comment (alicekwak) [reply automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/entities.yml:85]: Hmm, I don't think we need an explicit rule for lexicons (i.e., entity = Model), as the entities found by lexicons are automatically captured/labeled as Model.

Comment (alicekwak) [reply automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/entities.yml:85]: Also, could you explain what you meant by grouping the rules? Did you mean merging similar rules?

File: automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml
Comment (alicekwak) [reply automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml:36]: I switched the labels for those "model cannot" rules to ModelLimitation. It seems to work fine, so I'll keep it this way. If there's any issues later, we can try attachment then.

File: automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml
Comment (alicekwak) [reply automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml:155]: Right, I was thinking about that too. I'll try writing some tests for model descriptions today so that we can discuss this tomorrow!

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala
Comment (alicekwak) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala:385]: Thanks for the advice! I switched `Set.empty` to `origModel.attachments`.

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceMention.scala
Comment (alicekwak) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceMention.scala:25]: As you expected, the arg texts show up correctly, but the full mention span is off. I don't think that causes any major problems (at least for now), but it would be better if we could capture the full mention span correctly as well. Also, I need some help with serializing the mention, so let's definitely talk about this tomorrow. Thank you!

File: automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/entities.yml
Comment (maxaalexeeva) [reply automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/entities.yml:85]: meeting notes: delete this rule; move models to a new file

File: automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml
Comment (maxaalexeeva) [reply automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml:168]: meeting note: add pronouns to the "our model" model rule; filter out model descriptions with pronouns if there is no model name to resolve coreference to

File: automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml
Comment (maxaalexeeva) [reply automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml:155]: Is this resolved?

File: automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml
Comment (maxaalexeeva) [reply automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml:168]: resolved?

File: automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/vars.yml
Comment (maxaalexeeva) [inline automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/vars.yml:51]: nice!

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala
Comment (maxaalexeeva) [inline automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala:368]: does this cover "It"?

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala
Comment (maxaalexeeva) [inline automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala:1102]: Might be able to do Seq.empty instead of Seq().

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/actions/ExpansionHandler.scala
Comment (maxaalexeeva) [inline automates/text_reading/src/main/scala/org/clulab/aske/automates/actions/ExpansionHandler.scala:608]: do you think it's safe to remove the commented out relations? just to clean up?

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/apps/ExtractAndExport.scala
Comment (maxaalexeeva) [inline automates/text_reading/src/main/scala/org/clulab/aske/automates/apps/ExtractAndExport.scala:37]: Yes! Thanks!

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceEventMention.scala
Comment (maxaalexeeva) [inline automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceEventMention.scala:38]: Did you want to make a new text method for this type of event that will make a text span from arguments? Or you c could just borrow the one from Eidos (where this class came from---let me know if you need help refinding it). The generic "text" method will probably not get a good cross-sentence text. I can help with this (should be fast for me, up to you)

File: automates/text_reading/src/test/scala/org/clulab/aske/automates/text/TestModelDescrs.scala
Comment (maxaalexeeva) [inline automates/text_reading/src/test/scala/org/clulab/aske/automates/text/TestModelDescrs.scala:7]: I got rid of 'the' in the phrase because that's what we do with most other events. I wonder if there is a way to apply coref resolver here as well? Like if model name contains a certain list of generic words like simple, big, small, and such, resolve it to a model name within the sentence. Let'd discuss tomorrow. Fine for this pr. Also might be fine to include that extractions in the test.

**Review by maxaalexeeva — CHANGES_REQUESTED** [review]
A lot of cool stuff! Some comments are just for you to confirm things have been resolved. If they are, you can mark conversations resolved.

File: automates/text_reading/src/main/resources/Model.tsv
Comment (alicekwak) [reply automates/text_reading/src/main/resources/Model.tsv:30]: Thanks!

File: automates/text_reading/src/main/resources/application.conf
Comment (alicekwak) [reply automates/text_reading/src/main/resources/application.conf:65]: Yes, I've done changed Model to model (and added a new rule for capturing B-model) on my other branch, but I'll apply that to this branch as well! As for filtering out verb flee, I'm not sure if it'll be okay to do that. I noticed when model names do not contain any capital letters (i.e., `flee`), they're always captured as VB no matter what.

File: automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/context.yml
Comment (alicekwak) [reply automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/context.yml:61]: I didn't move this to unused rule because I was planning to fix and use that in the future. So I thought it'd be better to just temporarily disable.

File: automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/entities/grammar/avoid.yml
Comment (alicekwak) [reply automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/entities/grammar/avoid.yml:15]: Yes I agree. I'll keep this as disabled here with the note.

File: automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/functions.yml
Comment (alicekwak) [reply automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/functions.yml:186]: I don't want to get rid of this rule, though it malfunctions for now. Would it be better to move this rule to unused rule? I'm planning to fix this and use it later, though.

File: automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml
Comment (alicekwak) [reply automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml:155]: Yes, I wrote some model description tests under "TestModelDescrs", and so is resolved :)

File: automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml
Comment (alicekwak) [reply automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/modelDescriptions.yml:168]: Yes, pronouns are now captured by "model_pronouns" rule. They are used for model description rules, and then filtered out when all the extractions are done. So this is resolved!

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceMention.scala
Comment (alicekwak) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceMention.scala]: I guess this is resolved now, with all you helps! Thanks again, Masha :)

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala
Comment (alicekwak) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala:368]: Yes, "It" (or "it") is captured by "model_pronouns" rule.

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala
Comment (alicekwak) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala:1102]: Is it better to use Seq.empty instead of Seq()? What's the difference between the two?

Comment (alicekwak) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala:1102]: I changed the ones (`Seq()`) in combineFunction to Seq.empty, but I noticed there are more `Seq()` in other methods - Would it be better to replace them all with Seq.empty?

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/actions/ExpansionHandler.scala
Comment (alicekwak) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/actions/ExpansionHandler.scala:608]: Yes, I don't see any harm. I'll clean up those in this branch. Also, I've noticed some disabled lines in ExpansionHandler - would it be okay to clean-up those too? (see below for examples)

`//    println(s"$proportion --> ${mention.sentenceObj.getSentenceText}"`
`//          println("expanded arg: " + expandedWithIdentifierLabelDropped.text + " " + expandedWithIdentifierLabelDropped.foundBy + " " + expandedWithIdentifierLabelDropped.label + "|" + expandedWithIdentifierLabelDropped.labels.mkString("::"))`
`          //.map(addSubsumedAttachments(_, state))
          //.map(attachDCT(_, state))
          //.map(addOverlappingAttachmentsTextBounds(_, state))`
`//println(s"orig: ${orig.text}\texpanded: ${expanded.text}")`
`//    println("\noriginal:  " + entity.text + " " + entity.labels.mkString("") + " " + entity.foundBy)
//    println("expanded:  " + incomingExpanded.text + " " + incomingExpanded.labels.mkString("") ++ incomingExpanded.foundBy)` and a few more like these (+ is "attachments helper methods" on line 385 needed here?)

File: automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/functions.yml
Comment (maxaalexeeva) [reply automates/text_reading/src/main/resources/org/clulab/aske_automates/grammars/functions.yml:186]: That works!

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala
Comment (maxaalexeeva) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala:1102]: Yep

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/actions/ExpansionHandler.scala
Comment (maxaalexeeva) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/actions/ExpansionHandler.scala:608]: keep printouts---I use them for debugging. yesterday I thought I'd had to write printouts again but luckily they were already there :D

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceEventMention.scala
Comment (alicekwak) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceEventMention.scala:38]: It would be great if it's possible to get a text span from arguments, but they are in two different sentences... Would it be possible to get a text span over two sentences?

Comment (maxaalexeeva) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceEventMention.scala:38]: The args should have the information about the correct sentence. I don't think it should be a problem.

File: automates/text_reading/src/test/scala/org/clulab/aske/automates/text/TestModelDescrs.scala
Comment (alicekwak) [reply automates/text_reading/src/test/scala/org/clulab/aske/automates/text/TestModelDescrs.scala:7]: Yes, determiners are not needed here. Thank you!

Comment (alicekwak) [reply automates/text_reading/src/test/scala/org/clulab/aske/automates/text/TestModelDescrs.scala:7]: And coref resolver might be applicable here, I'm not sure though. But that is an interesting idea! Let's discuss this in our meeting tomorrow.

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/actions/ExpansionHandler.scala
Comment (alicekwak) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/actions/ExpansionHandler.scala:608]: Okay, and "attachments helper methods" are also needed there?

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceEventMention.scala
Comment (alicekwak) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceEventMention.scala:38]: You're right. Can we discuss this at tomorrow's meeting?

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/actions/ExpansionHandler.scala
Comment (maxaalexeeva) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/actions/ExpansionHandler.scala]: yeah you can get rid of those helper methods. our attachments are different than what they do in the project that was used as a template for our system.

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala
Comment (maxaalexeeva) [inline automates/text_reading/src/main/scala/org/clulab/aske/automates/OdinActions.scala:378]: Sorry, I should have caught this sooner, but below are a few comments about this method and the next two. You can try to adjust this based on the comments (we can meet to discuss) OR I can most likely get this done today or tomorrow. 

1) I think there might be a way to simplify this. You can have your previous model method return the mention you think is the previous Model mention----if you return that, that will have the info on both interval and sentence, so you won't have to have a separate sentence method in addition to the return interval method. This way you also won't have to construct a new text bound mention in line 379. To make sure you can account for cases where no previous model being available, this method can return an option (Option[Mention]) and not just a mention. 
2) I think it is possible for a previous model to not necessarily be in the previous sentence---it can easily be within the same sentence. I think we can just get all the model mentions available in the current and previous sentence and choose, let's say, the latest one of them.

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceEventMention.scala
Comment (maxaalexeeva) [inline automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceEventMention.scala:75]: About the note: both you and I need to read a bit on classes vs objects in Scala to answer this :) Either case, for now, keep the object and delete the note

**Review by maxaalexeeva — CHANGES_REQUESTED** [review]
@alicekwak, most of this looks good to me! There is one major-ish change (and a couple of minor things) that might be useful. Let me know if you'd like me to address it myself or if you want to discuss.

**Review by maxaalexeeva — DISMISSED** [review]

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceEventMention.scala
Comment (maxaalexeeva) [inline automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceEventMention.scala:36]: @alicekwak, the changes look good! A quick question: what is this method for?

Comment (maxaalexeeva) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceEventMention.scala:36]: Never mind. Looks like a method that you can use to copy a mention with some modifications.

**Review by maxaalexeeva — APPROVED** [review]
Looks good! Thanks!

File: automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceEventMention.scala
Comment (alicekwak) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceEventMention.scala:36]: That's correct!

Comment (cl4yton) [reply automates/text_reading/src/main/scala/org/clulab/aske/automates/mentions/CrossSentenceEventMention.scala:36]: @alicekwak : Can you please add an explicit note to capture the descriptive summary @maxaalexeeva gave?

**Review by cl4yton — APPROVED** [review]
Hi @alicekwak and @maxaalexeeva : Thank you for *all* of the work on this -- there is quite a lot. I cannot speak to the details, unfortunately, but I did read through everything and tried to follow what I can. T

he main question that this raises is: Do we have an "entry point" to how someone new coming onto this project can get to understanding the overall structure of all of the rules? I do like the detailed comments that explain how some rules are being suppressed b/c of interactions, that is helpful. But I'm not sure how to see the "big picture".

I suggest added one comment to include the summary interpretation that Masha provided.

Other than that, I think this is good and I approve.

Thanks!

**Review by maxaalexeeva — APPROVED** [review]
@alicekwak, looks good! Let me know if you need me to re-approve after you added the comment @cl4yton requested!

**Comment by alicekwak** [comment]
> Hi @alicekwak and @maxaalexeeva : Thank you for _all_ of the work on this -- there is quite a lot. I cannot speak to the details, unfortunately, but I did read through everything and tried to follow what I can. T
> 
> he main question that this raises is: Do we have an "entry point" to how someone new coming onto this project can get to understanding the overall structure of all of the rules? I do like the detailed comments that explain how some rules are being suppressed b/c of interactions, that is helpful. But I'm not sure how to see the "big picture".
> 
> I suggest added one comment to include the summary interpretation that Masha provided.
> 
> Other than that, I think this is good and I approve.
> 
> Thanks!

Hi @cl4yton, thank you for the review! I'll add the comment as a note as you suggested. And as for your question, I don't think we have an "entry point" to the rules for now, except for the Odin manual. We might need to write a manual for our own TR rules and methods (maybe at the wiki page?) that can work as an entry point to someone newly joining our project. I think that will help them grasp the big picture of how the rules / methods work.

**Comment by maxaalexeeva** [comment]
@cl4yton if by entry points you mean methods to run the rules, there are a few places I can point you to. I'll try to start updating the wiki this week with the TR developments from the past year. @alicekwak I might request some write-ups from you as well for the stuff you've been working on.

**Comment by maxaalexeeva** [comment]
Ah I see @alicekwak also mentioned the wiki. Yeah, we should start updating that.

**Review by maxaalexeeva — APPROVED** [review]
Looks good! Thanks!

**Comment by alicekwak** [comment]
> @cl4yton if by entry points you mean methods to run the rules, there are a few places I can point you to. I'll try to start updating the wiki this week with the TR developments from the past year. @alicekwak I might request some write-ups from you as well for the stuff you've been working on.

Sure! Just let me know which parts I should work on. I can write some explanations on function & context & model-related extractions (+ related methods).

**Pull request merged**
